File: cortex/enterprise/sources.py
from __future__ import annotations

from dataclasses import dataclass

from cortex.episodic.memory_store import EpisodicMemoryStore
from cortex.models import EpisodicHit, SemanticDocument
from cortex.semantic.vault_reader import VaultReader


@dataclass(frozen=True)
class VaultSource:
    path: str
    scope: str
    project_id: str


@dataclass(frozen=True)
class EpisodicSource:
    persist_dir: str
    scope: str
    project_id: str
    collection_name: str


class MultiVaultReader:
    def __init__(self, *, sources: list[VaultSource], embedding_model: str, embedding_backend: str) -> None:
        self.sources = sources
        self._readers: list[tuple[VaultSource, VaultReader]] = [
            (
                source,
                VaultReader(
                    vault_path=source.path,
                    embedding_model=embedding_model,
                    embedding_backend=embedding_backend,
                ),
            )
            for source in sources
        ]

    def search(self, query: str, top_k: int, use_embeddings: bool = True) -> list[SemanticDocument]:
        hits: list[SemanticDocument] = []
        for source, reader in self._readers:
            source_hits = reader.search(query, top_k=top_k, use_embeddings=use_embeddings)
            for doc in source_hits:
                hits.append(
                    doc.model_copy(
                        update={
                            "origin_scope": source.scope,
                            "origin_project_id": source.project_id,
                            "origin_vault": source.path,
                            "origin_persist_dir": "",
                        }
                    )
                )
        return hits


class MultiEpisodicReader:
    def __init__(self, *, sources: list[EpisodicSource], embedding_model: str, embedding_backend: str) -> None:
        self.sources = sources
        self._stores: list[tuple[EpisodicSource, EpisodicMemoryStore]] = [
            (
                source,
                EpisodicMemoryStore(
                    persist_dir=source.persist_dir,
                    embedding_model=embedding_model,
                    embedding_backend=embedding_backend,
                    collection_name=source.collection_name,
                ),
            )
            for source in sources
        ]

    def search(self, query: str, top_k: int, use_embeddings: bool = True) -> list[EpisodicHit]:
        hits: list[EpisodicHit] = []
        for source, store in self._stores:
            source_hits = store.search(query, top_k=top_k, use_embeddings=use_embeddings)
            for hit in source_hits:
                metadata = dict(hit.entry.metadata)
                metadata.update(
                    {
                        "scope": source.scope,
                        "project_id": source.project_id,
                        "origin_vault": "",
                        "origin_persist_dir": source.persist_dir,
                    }
                )
                entry = hit.entry.model_copy(update={"metadata": metadata})
                hits.append(
                    hit.model_copy(
                        update={
                            "entry": entry,
                            "origin_scope": source.scope,
                            "origin_project_id": source.project_id,
                            "origin_vault": "",
                            "origin_persist_dir": source.persist_dir,
                        }
                    )
                )
        return hits

File: rust/crates/cortex-enterprise/src/sources.rs
//! Puerto de `cortex.enterprise.sources`: lectores multi-vault y
//! multi-episódico. Los tipos son owned y anotan origen (scope/project/
//! vault/persist_dir) exactamente como los `model_copy(update=…)` de Python.

use crate::error::EnterpriseError;
use cortex_app::episodic::MemoryEntry;

/// Scope de origen de un hit.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum SourceScope {
    Local,
    Enterprise,
}

impl SourceScope {
    pub fn as_str(self) -> &'static str {
        match self {
            Self::Local => "local",
            Self::Enterprise => "enterprise",
        }
    }
}

/// Espejo de `VaultSource`.
#[derive(Debug, Clone)]
pub struct VaultSource {
    pub path: String,
    pub scope: SourceScope,
    pub project_id: String,
}

/// Espejo de `EpisodicSource`.
#[derive(Debug, Clone)]
pub struct EpisodicSource {
    pub persist_dir: String,
    pub scope: SourceScope,
    pub project_id: String,
    pub collection_name: String,
}

/// SemanticDocument anotado con origen.
#[derive(Debug, Clone)]
pub struct SemanticHit {
    pub path: String,
    pub title: String,
    pub content: String,
    pub score: f64,
    pub origin_scope: SourceScope,
    pub origin_project_id: String,
    pub origin_vault: String,
    pub origin_persist_dir: String,
}

impl SemanticHit {
    pub fn new(
        path: impl Into<String>,
        title: impl Into<String>,
        content: impl Into<String>,
        score: f64,
    ) -> Self {
        Self {
            path: path.into(),
            title: title.into(),
            content: content.into(),
            score,
            origin_scope: SourceScope::Local,
            origin_project_id: String::new(),
            origin_vault: String::new(),
            origin_persist_dir: String::new(),
        }
    }

    /// Anota origen (equivalente al `model_copy(update={origin_*})`).
    pub fn with_origin(mut self, scope: SourceScope, project_id: String, vault: String) -> Self {
        self.origin_scope = scope;
        self.origin_project_id = project_id;
        self.origin_vault = vault;
        self
    }
}

/// `EpisodicHit` con entry + origen.
#[derive(Debug, Clone)]
pub struct EpisodicHit {
    pub entry: MemoryEntry,
    pub score: f64,
    pub origin_scope: SourceScope,
    pub origin_project_id: String,
    pub origin_vault: String,
    pub origin_persist_dir: String,
}

/// Backend de búsqueda inyectable por fuente (seam de testabilidad; el
/// adapter nativo vive en `retrieval::native`).
pub trait SearchBackend: Send {
    fn search_vault(
        &mut self,
        source: &VaultSource,
        query: &str,
        top_k: usize,
        use_embeddings: bool,
    ) -> Result<Vec<SemanticHit>, EnterpriseError>;
    fn search_episodic(
        &mut self,
        source: &EpisodicSource,
        query: &str,
        top_k: usize,
        use_embeddings: bool,
    ) -> Result<Vec<EpisodicHit>, EnterpriseError>;
}
